src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -664,38 +664,36 @@ protected internal void OnAnalyzerReferenceRemoved(ProjectId projectId, Analyzer
             }
         }
 
-        private static ImmutableHashSet<string> GetFilePaths(ImmutableArray<DocumentInfo> documents)
-        {
-            var map = ImmutableHashSet<string>.Empty.WithComparer(StringComparer.OrdinalIgnoreCase);
-            foreach (var document in documents)
-            {
-                map = map.Add(document.FilePath);
-            }
-            return map;
-        }
-
         /// <summary>
         /// Call this method when a document is added to a project in the host environment.
         /// </summary>
         protected internal void OnDocumentAdded(DocumentInfo documentInfo)
         {
             using (_serializationLock.DisposableWait())
             {
-                OnDocumentAdded_NoLock(documentInfo);
+                var documentId = documentInfo.Id;
+
+                var oldSolution = this.CurrentSolution;
+                var newSolution = this.SetCurrentSolution(oldSolution.AddDocument(documentInfo));
+
+                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentAdded, oldSolution, newSolution, documentId: documentId);
```

Comment:
Not changed in this PR, but is it a good idea to raise the event while holding a lock?

Reply:
The method isn't well named: it only adds the event to a queue, and ultimately isn't ran when the lock is held. But good question.